bm_serial.c

```diff
@@ -691,6 +691,119 @@ bm_serial_error_e bm_serial_cfg_delete_response(uint64_t node_id, bm_common_conf
+  bm_serial_error_e rval = BM_SERIAL_OK;
+  do {
+    uint16_t message_len = sizeof(bm_serial_packet_t) + sizeof(bm_serial_device_info_reply_t) + bcmp_info->dev_name_len + bcmp_info->ver_str_len;
+    bm_serial_packet_t *packet = _bm_serial_get_packet(BM_SERIAL_DEVICE_INFO_REPLY, 0, message_len);
+
+    if(!packet) {
+      rval = BM_SERIAL_OUT_OF_MEMORY;
+      break;
+    }
+    bm_serial_device_info_reply_t *device_info_reply_msg = (bm_serial_device_info_reply_t *)packet->payload;
+    memcpy(device_info_reply_msg, bcmp_info, sizeof(bm_serial_device_info_reply_t)+ bcmp_info->dev_name_len + bcmp_info->ver_str_len);
+    packet->crc16 = bm_serial_crc16_ccitt(0, (uint8_t *)packet, message_len);
+    if(!_callbacks.tx_fn((uint8_t *)packet, message_len)) {
+      rval = BM_SERIAL_TX_ERR;
+      break;
+    }
+  } while(0);
+  return rval;
+}
+
+bm_serial_error_e bm_serial_send_resource_request(uint64_t node_id){
+  bm_serial_error_e rval = BM_SERIAL_OK;
+  do {
+    uint16_t message_len = sizeof(bm_serial_packet_t) + sizeof(bm_serial_resource_table_request_t);
+    bm_serial_packet_t *packet = _bm_serial_get_packet(BM_SERIAL_RESOURCE_REQ, 0, message_len);
+
+    if(!packet) {
+      rval = BM_SERIAL_OUT_OF_MEMORY;
+      break;
+    }
+    bm_serial_resource_table_request_t *resource_req_msg = (bm_serial_resource_table_request_t *)packet->payload;
+    resource_req_msg->target_node_id = node_id;
+    packet->crc16 = bm_serial_crc16_ccitt(0, (uint8_t *)packet, message_len);
+    if(!_callbacks.tx_fn((uint8_t *)packet, message_len)) {
+      rval = BM_SERIAL_TX_ERR;
+      break;
+    }
+  } while(0);
+  return rval;
+}
+
+// TODO - This struct is defined in bcmp_messages.h, but that is not included in this repo yet.
+// We should eventually include it, or have this struct moved into the common messages repo.
+// Perhaps this can happen when we do this ticket: SC-195824
+typedef struct {
+  // Length of resource name
+  uint16_t resource_len;
+  // Name of resource
+  char resource[0];
+} __attribute__((packed)) bcmp_resource_t;
+
+bm_serial_error_e bm_serial_send_resource_reply(uint64_t node_id, bm_serial_resource_table_reply_t* bcmp_resource) {
+  ( void ) node_id;
+  bm_serial_error_e rval = BM_SERIAL_OK;
+  do {
+    uint16_t message_len = sizeof(bm_serial_packet_t) + sizeof(bm_serial_resource_table_reply_t);
+
+    size_t length_of_resources = 0;
+    uint16_t num_pubs = bcmp_resource->num_pubs;
+    while(num_pubs){
+        bcmp_resource_t * cur_resource = (bcmp_resource_t *)(&bcmp_resource->resource_list[length_of_resources]);
+        length_of_resources += (sizeof(bcmp_resource_t) + cur_resource->resource_len);
+        num_pubs--;
+    }
+    uint16_t num_subs = bcmp_resource->num_subs;
+    while(num_subs){
+        bcmp_resource_t * cur_resource = (bcmp_resource_t *)(&bcmp_resource->resource_list[length_of_resources]);
+        length_of_resources += (sizeof(bcmp_resource_t) + cur_resource->resource_len);
+        num_subs--;
+    }
```

Comment:
calculate the total length of the resources, which make up the zero length array of the bm_serial_resource_table_reply_t message.